### PR TITLE
feat: implement Client::updateGroup()

### PR DESCRIPTION
**Pull request by willsantanentcs**
If approved, this PR will implement the updateGroup() method in Client.php. Part 5/5 for iteration 2.

### CODE REVIEW

**Review by tomegantcs — COMMENTED** [review]
Does it work? Maybe. Is it usable? I don't think so. Here's the conceptual problem in object oriented programming I think of an object modeling a thing. That's why we give them noun names typically. They stand in for something in the real world though so they are more like pronouns though. I could be modeled by a User object for instance. I can do things like run, talk, and sleep. But the action is not something I can do it is something the API can do. Therefore it feels weird, there is a sort of cognitive dissonance in using the Client because the user objects are not people, groups are not collections of people, learning Modules are not content on the SmarterU platform but instead they are API queries. Sure we could call them UserQuery, GroupQuery etc. and that would I think help but as a library user what I want is to be able to ask "higher order" questions like Hey API who are the users in this organization? Hey API add Bob to the Employee group. Which to me means I want the client to have methods like `addUser(<User>)`, `listUsersInGroup(<Group|string groupname>): <List<User>>` I don't want to built query objects I want that to be the responsibility of the Client since I don't know enough about the SmarterU API to get it right. I want to pack only the data describing the user in the User object and I don't want to have to know the magic strings to dereference some attribute of the user from a hash... (this is one of the things that makes typical javascript and python painful to use).

So I'm going to wait and see what the documentation looks like but I think we ought to get back together as a team with Jacob and Brian and change this design to more closely match that specified in the design doc.

**Comment by jacobalvareztcs** [comment]
I agree with Tom. Although, isn't `updateGroup` here kinda ok in that regard since we're passing a `Group` and the `Client` is handling the XML stuff?

I'm not a fan of returning an array with the response and errors. Seems like we should return a `Response` class that abstracts that away, since all API calls return an "API Response XML Package". Also, if there are errors then doesn't that mean the call failed and we should throw an exception, not return them?

**Comment by tomegantcs** [comment]
> I agree with Tom. Although, isn't `updateGroup` here kinda ok in that regard since we're passing a `Group` and the `Client` is handling the XML stuff?

I think Group actually handles the XML. And that's my objection. By focusing on serializing to XML, Group is less the group and more a Request.
 
> I'm not a fan of returning an array with the response and errors. Seems like we should return a `Response` class that abstracts that away, since all API calls return an "API Response XML Package". Also, if there are errors then doesn't that mean the call failed and we should throw an exception, not return them?

I think Will has found that not all errors are failures so no not all errors rise to be exceptions. As a rule fatal errors should be exceptions but non fatal errors... I like Jacob's thought about returning a Response. The Response ought to have a status of some sort success/failure. If failure it ought to include the issues. But if successful (and in the case of this API if partially successful) it ought to let me get the value... `UserListResponse->getValue()` would return `array<User>` for instance.

**Comment by jacobalvareztcs** [comment]
> I think Group actually handles the XML. And that's my objection. By focusing on serializing to XML, Group is less the group and more a Request.

Ah I see now

**Comment by willsantanentcs** [comment]
Marked as draft for now because I discovered that the XML isn't quite right while I was working on the documentation.

File: src/DataTypes/Group.php [hunk 1]
```diff
@@ -679,17 +679,56 @@ public function toXML(
...
+                $notificationEmails->addChild('NotificationEmail', $email);
+            }
         }
-        $group->addChild('Status', $this->getStatus());
-        $group->addChild('Description', $this->getDescription());
-        $group->addChild('HomeGroupMessage', $this->getHomeGroupMessage());
-        $notificationEmails = $group->addChild('NotificationEmails');
-        foreach ($this->getNotificationEmails() as $email) {
-            $notificationEmails->addChild('NotificationEmail', $email);
+        else {
```
Comment (tomegantcs) [inline src/DataTypes/Group.php:696]: NIT typically we cuddle else to the close brace of an if or if-else block

File: tests/Client/UpdateGroupClientTest.php [hunk 2]
```diff
@@ -574,45 +567,29 @@ public function testUpdateGroupProducesExpectedInputWithOnlyRequiredInfo() {
         foreach ($packageAsXml->Parameters->children() as $parameter) {
             $parameters[] = $parameter->getName();
         }
-        self::assertCount(1, $parameters);
+        self::assertCount(2, $parameters);
+        self::assertContains('Identifier', $parameters);
+        $identifierTags = [];
+        foreach ($packageAsXml->Parameters->Identifier->children() as $tag) {
```
Comment (tomegantcs) [inline tests/Client/UpdateGroupClientTest.php:573]: NIT let's add to our minimum cleanup sprint changing these `packageAsXml` to `bodyAsXml` to make these test more readable.

**Review by tomegantcs — APPROVED** [review]

**Pull request merged**